README.md
### [**SCANsat**][top]: Real Scanning, Real Science, Warp Speed!
[![][shield:support-ksp]][KSP:developers]&nbsp;
[![][shield:jenkins-rel]][SCANsat:rel-jenkins]&nbsp;
[![][shield:license-bsd]][SCANsat:rel-license]&nbsp;
![scan your planetoid like the big boys do][bigmap-scan-10000x]
> ###### **Example SAR scan of Kerbin at 1000x and then 10,000x warp**

[![][shield:support-rpm]][RPM:release]&nbsp;
[![][shield:support-ket]][Kethane:release]&nbsp;
[![][shield:support-ors]][ORS:release]&nbsp;
[![][shield:support-mm]][mm:release]&nbsp;
[![][shield:support-toolbar]][toolbar:release]&nbsp;
[![][shield:support-kspi]][kspi:release]&nbsp;
[![][shield:support-karbonite]][karbonite:release]&nbsp;
[![][shield:support-usi]][usi:release]&nbsp;
[![][shield:support-epl]][epl:release]&nbsp;

**Table of Contents**
------------------------------------------
* [0. People, Facts, and FAQs][0]
  * [a. Facts][0a]
  * [b. FAQs][0b]
* [1. Installation and Interoperability][1]
  * [a. Installation][1a]
  * [b. GameData Layout][1b]
  * [c. Other Add-Ons][1c]
* [2. Types of Scans][2]
  * [a. Natvive SCANsat][2a]
  * [b. Resource Scans][2b]
  * [1. Kethane][2b1] [![][shield:support-ket]][kethane:patch01][![][shield:jenkins-ket]][SCANsat:ket-jenkins]
  * [2. ORS][2b2] [![][shield:support-ors]][ORS:release][![][shield:jenkins-ors]][SCANsat:ors-jenkins]
* [3. Basic Usage][3]
  * [a. FAQ: Finding a Good Altitude][3a]
  * [b. Mismatched Scanners][3b]
* [4. Big Map][4]
* [5. Parts and Sensors Types][5]
  * [a. RADAR][5a]
  * [b. SAR][5b]
  * [c. Multi][5c]
  * [d. BTDT][5d]
  * [e. MapTraq (deprecated)][5e]
* [6. (Career Mode) Research and Development][6]
  * [a. Minimum Scan for Science (30%)][6a]
  * [b. Getting Maximum Science][6b]
* [7. Background Scanning][7]
* [8. Time Warp][8]
* [9. Note: Data Sources][9]

**WARNING**:

This add-on is a **work**-in-**progress**.

This means you should expect that it **may not work**, and you should be unsurprised if it **does not progress**.

Disclaimer aside, this add-on is widely used and it **usually** works just fine.

### [:top:][top] 0. People, Facts, and FAQs
------------------------------------------
#### Maintainers 
The current maintainer is:
  + [technogeeky][technogeeky] \<<technogeeky@gmail.com>\>

Maintainers are the people who you should complain to if there is something wrong.

Complaints in various forms are prioritized as follows:

  1. [Pull Requests][SCANsat:pulls] are given the **highest** priority possible. ~ 24 hour response
  2. [Issues][SCANsat:issues] are given *higher* priority than other complaints. ~ 2 day response
  3. [E-Mails][SCANsat:email] will be answered as soon as possible (it's a forwarded list) ~ 3 day response
  4. [Forum Posts][SCANsat:rel-thread] are given a medium priority. ~ 1 week response
  5. [Forum Private Messages](http://forum.kerbalspaceprogram.com/private.php) are given a low priority. We might forget!
  6. [Reddit Posts and PMs][KSP:reddit] are the lowest priority. We often lurk and don't login!

If you submit a well-reasoned pull request, you may even trigger a new release!

#### Authors
The current authors include:
  + [technogeeky][technogeeky] \<<technogeeky@gmail.com>\>
  + [DMagic][DMagic] \<<david.grandy@gmail.com>\>

Past authors include:
  + [damny][damny] \<<missing-in-action@nowhere-to-be-found.com>\>

As of August 2014, the vast majority of code is damny's and DMagic's; and technogeeky and is slowly helping out here and there.

#### Contributors

In addition to the authors, the following people have contributed:
  + (Models, Graphics, Textures) [Milkshakefiend][Milkshakefiend]

#### Licenses

For licensing information, please see the [included LICENSE.txt][SCANsat:rel-license] file.

[Source Code][SCANsat:source] is available, as some licenses may require.

### [:top:][top] a. Facts


### [:top:][top] b. FAQs

  * What does SCANsat do?
    * It allows you to scan planetary bodies for terrain, biome, and resource information and generate various kinds of maps.
  * How does SCANsat affect gameplay?
    * It allows you to see surface details from orbit from an interactive, zoom-able map. This will help you plan your missions (for example, landing near a divider between two or three biomes) and provide critical information you need to attempt a safe landing (for instance, the slope map will help you avoid treacherous hills)
  * Will this version break my existing scans from older versions of SCANsat?
    * **No!** This version is completely backwards compatible, and you current scanning state (which is stored in persistent.sfs) will be safe and sound. Nevertheless, you should make a backup copy of your game before upgading any mod.
  * Do I need to attach a part to my vessel to use SCANsat?
    * **No, but...**. You can view existing maps from any vessel, but you need to attach a scanner to add new data to the maps.
  * [Career Mode] Does SCANsat give us science points?
    * **Yes!** For each type of map, if you scan at least 30% of the surface, you can yse Data for partial science points; up until the maximum value at 95% map coverage.
  * [Career Mode] Is it integrated into the tech tree?
    * **Yes!** This link tells you which nodes unlock which parts in the tech tree.
  * [Contracts] Does SCANsat offer contracts to complete?
    * **No.** This is a planned feature.
  * Can you add <some feature or change> to SCANsat?
    * **Probably!** First, check the issues page to see if it's already been requested. If not, add a new issue. Even better, attempt to add the feature yourself and submit a pull request. We'll catch the bugs for you!



### [:top:][top] 1. Installation and Interoperability 
------------------------------------------
:heavy_exclamation_mark:

#### [:top:][top] a. Installation 
:heavy_exclamation_mark:

  1. Put the SCANsat folder in your KSP installation's GameData folder.
  2. (Optional) Place the SCANsatRPM folder in your KSP installation's GameData folder.

#### [:top:][top] b. GameData Layout 
:heavy_exclamation_mark:

#### [:top:][top] c. Other Add-Ons 

S.C.A.N. is proud to collaborate with other KSP mods and modding teams. Following is a table of all of the mods, add-ons, or software that we interoperate with.

**Built Using** | **Supported By**
:---: | :---:
[![Support for Kethane][kethane:logo]][kethane:release] | [![Support for MKS][usi:logo]][usi:release]
[**OpenResourceSystem**][ors:release] | [![Support for ALCOR][alcor:logo]][alcor:release]
[**RasterPropMonitor**][rpm:release]  | [![Support for Karbonite][karbonite:logo]][karbonite:release]

SCANsat/CHANGELOG.txt
Build 8 - 2014-09-10
---------------------

- Updated for KSP 0.24.2

Resource Scanning:
- Planetary resource overlay for big map
	- Supports latest Kethane version - 0.9
	- Supports ORS resources with ORS v1.3
	- All resources types can be scanned in the background; no need to remain in control of a vessel while scanning
	- All resource scanning data is persistent; no different from standard SCANsat sensors
	- Resource types controlled through the SCANsat settings menu
	- Resources require explicit support from SCANsat
		- Users cannot add or change which resource types are supported
	
- Module Manager configs included for Kethane, Extraplanetary Launchpad and Interstellar resources
	- Support for Modular Kolonization System and Karbonite are provided in their respective packages
	
- ORS resources
	- Resources displayed on the big map with a place holder color overlay
	- ORS plugin not included with SCANsat package; compatible with any version of ORS included in other mods
	
- Kethane resources
	- Background scanning requires using the SCANsat modules added through the included Module Manager configs
		- This function is separate from standard Kethane scanning; both are accessed through the right-click context menu
	- SCANsat will read existing Kethane scanning data if present; no SCANsat scanners are required for this
	- Kethane grid can be updated through the SCANsat settings menu
		- Grid must be forced to reset before changes become visible; saving/loading or switching which resource type is displayed works
	- Kethane resources displayed with their specified color values based on deposit concentration

Parts:
- MapTraq deprecated
	- Part still exists but is not available in the VAB/SPH
	- Part functionality is completely removed
	- Module Manager configs for adding the MapTraq SCANsat module should be removed (having the module shouldn't hurt though)
	
- Scanning altitude indicator is present for all SCANsat scanners, including resource scanners
	- Displayed in the right-click context menu
	
- Improved Multi-Spectral scanner
	- Improved clickability
	- More efficient model; 60% reduction in triangle count
	
- Rescaled the BTDT to be much smaller

- Part cost balancing

BackEnd Changes:
- Background scanning will function in every scene where time passes
	- Flight, Map, Tracking Station and Space Center
	- Can be turned off from the SCANsat settings menu
- SCANsat parts not required for background scanning during flight
- SCANsat parts not required for toolbar icons and functional maps
	- Maps and menus don't work outside of the flight/map scenes
	
- Persistent scanning data storage has been altered
	- Requires a one-time, automatic conversion from the old to the new method
	- There have been no complaints as of yet, but users upgrading from SCANsat v6.x may want to make backups of any existing save files
	
- ModStatistics removed
	
Bug Fixes:
- Science can be gathered from all planets
	- Reduced science return for planets without terrain/biomes
- Active scanners no longer play through their animation on startup; they start fully deployed
- Toolbar icons fallback to place holder textures if the default SCANsat icons are moved/altered/deleted

SCANsat/SCANcontroller.cs
/* 
 * [Scientific Committee on Advanced Navigation]
 * 			S.C.A.N. Satellite
 *
 * SCANcontroller - scenario module that handles all scanning
 * 
 * Copyright (c)2013 damny;
 * Copyright (c)2014 David Grandy <david.grandy@gmail.com>;
 * Copyright (c)2014 technogeeky <technogeeky@gmail.com>;
 * Copyright (c)2014 (Your Name Here) <your email here>; see LICENSE.txt for licensing details.
 */

using System;
using System.Linq;
using System.Collections.Generic;
using UnityEngine;
using palette = SCANsat.SCANpalette;

namespace SCANsat
{
	public class SCANcontroller : ScenarioModule
	{
		public static SCANcontroller controller {
			get {
				Game g = HighLogic.CurrentGame;
				if(g == null) return null;
				foreach(ProtoScenarioModule mod in g.scenarios) {
					if(mod.moduleName == typeof(SCANcontroller).Name) {
						var tScene = mod.targetScenes.FirstOrDefault(t => t == GameScenes.SPACECENTER);
						if (tScene != GameScenes.SPACECENTER) {
							SCANUtil.SCANlog("Adding new target scenes to scenario module");
							mod.targetScenes.Add(GameScenes.SPACECENTER);
							mod.targetScenes.Add(GameScenes.TRACKSTATION);
						}
						return (SCANcontroller)mod.moduleRef;
					}
				}
				return (SCANcontroller)g.AddProtoScenarioModule(typeof(SCANcontroller), GameScenes.FLIGHT, GameScenes.SPACECENTER, GameScenes.TRACKSTATION).moduleRef;
			}
			private set { }
		}
		public static int minScanAlt = 5000;
		public static int maxScanAlt = 500000;
		[KSPField(isPersistant = true)]
		public int colours = 0;
		[KSPField(isPersistant = true)]
		public bool map_markers = true;
		[KSPField(isPersistant = true)]
		public bool map_flags = true;
		[KSPField(isPersistant = true)]
		public bool map_orbit = true;
		[KSPField(isPersistant = true)]
		public bool map_asteroids = true;
		[KSPField(isPersistant = true)]
		public bool map_grid = true;
        [KSPField(isPersistant = true)]
        public bool map_ResourceOverlay = false; //Is the overlay activated for the selected resource
		[KSPField(isPersistant = true)]
		public int projection = 0;
		[KSPField(isPersistant = true)]
		public int map_width = 0;
		[KSPField(isPersistant = true)]
		public int map_x = 100;
		[KSPField(isPersistant = true)]
		public int map_y = 50;
		[KSPField(isPersistant = true)]
		public string anomalyMarker = "✗";
		[KSPField(isPersistant = true)]
		public string closeBox = "✖";
		[KSPField(isPersistant = true)]
		public bool legend = false;
		[KSPField(isPersistant = true)]
		public bool scan_background = true;
		[KSPField(isPersistant = true)]
		public int timeWarpResolution = 20;
        [KSPField(isPersistant = true)]
        public bool globalOverlay = false; //Global resource overlay toggle
        [KSPField(isPersistant = true)]
        public int gridSelection = 0; //Which resource type is selected in the settings menu
        [KSPField(isPersistant = true)]
        public int resourceOverlayType = 0; //0 for ORS, 1 for Kethane
        [KSPField(isPersistant = true)]
        public bool dataRebuild = true;


		public static List<SCANdata.SCANResource> ResourcesList = new List<SCANdata.SCANResource>(); //The list of all relevant resources
		internal static Dictionary<string, SCANdata> body_data = new Dictionary<string, SCANdata>();

		public bool kethaneRebuild = false; //these three used by the kethane watcher
        public bool kethaneReset = false;
        public bool kethaneBusy = false;

		public override void OnLoad(ConfigNode node) {
			body_data.Clear();
			ResourcesList.Clear();
			ConfigNode node_vessels = node.GetNode("Scanners");
			if(node_vessels != null) {
				print("SCANsat Controller: Loading " + node_vessels.CountNodes.ToString() + " known vessels");
				foreach(ConfigNode node_vessel in node_vessels.GetNodes("Vessel")) {
					Guid id = new Guid(node_vessel.GetValue("guid"));
					foreach(ConfigNode node_sensor in node_vessel.GetNodes("Sensor")) {
						int sensor = Convert.ToInt32(node_sensor.GetValue("type"));
						double fov = Convert.ToDouble(node_sensor.GetValue("fov"));
						double min_alt = Convert.ToDouble(node_sensor.GetValue("min_alt"));
						double max_alt = Convert.ToDouble(node_sensor.GetValue("max_alt"));
						double best_alt = Convert.ToDouble(node_sensor.GetValue("best_alt"));
						registerSensor(id, (SCANdata.SCANtype)sensor, fov, min_alt, max_alt, best_alt);
					}
				}
			}
			ConfigNode node_progress = node.GetNode("Progress");
			if(node_progress != null) {
				foreach(ConfigNode node_body in node_progress.GetNodes("Body")) {
					string body_name = node_body.GetValue("Name");
					print("SCANsat Controller: Loading map for " + body_name);
					CelestialBody body = FlightGlobals.Bodies.FirstOrDefault(b => b.name == body_name);
					if (body != null) {
						SCANdata data = SCANUtil.getData(body);
						try {
							string mapdata = node_body.GetValue("Map");
							if (dataRebuild) { //On the first load deserialize the "Map" value to both coverage arrays
								SCANUtil.integerDeserialize(mapdata, true, data);
								//data.deserialize(mapdata);
							}
							else {
								SCANUtil.integerDeserialize(mapdata, false, data);
							}
						} catch(Exception e) {
							print(e.ToString());
							print(e.StackTrace);
							// fail somewhat gracefully; don't make the save unloadable 
						}
						data.disabled = Convert.ToBoolean(node_body.GetValue("Disabled"));
					}
				}
			}
			dataRebuild = false; //Used for the one-time update to the new integer array
			if (HighLogic.LoadedScene == GameScenes.FLIGHT)
				Resources(FlightGlobals.currentMainBody);
		}

		public override void OnSave(ConfigNode node) {
			ConfigNode node_vessels = new ConfigNode("Scanners");
			foreach(Guid id in knownVessels.Keys) {
				ConfigNode node_vessel = new ConfigNode("Vessel");
				node_vessel.AddValue("guid", id.ToString());
				if(knownVessels[id].vessel != null) node_vessel.AddValue("name", knownVessels[id].vessel.vesselName); // not read
				foreach(SCANsensor sensor in knownVessels[id].sensors.Values) {
					ConfigNode node_sensor = new ConfigNode("Sensor");
					node_sensor.AddValue("type", (int)sensor.sensor);
					node_sensor.AddValue("fov", sensor.fov);
					node_sensor.AddValue("min_alt", sensor.min_alt);
					node_sensor.AddValue("max_alt", sensor.max_alt);
					node_sensor.AddValue("best_alt", sensor.best_alt);
					node_vessel.AddNode(node_sensor);
				}
				node_vessels.AddNode(node_vessel);
			}
			node.AddNode(node_vessels);
			ConfigNode node_progress = new ConfigNode("Progress");
			foreach (string body_name in SCANcontroller.body_data.Keys)
			{
				ConfigNode node_body = new ConfigNode("Body");
				SCANdata body_scan = SCANcontroller.body_data[body_name];
				node_body.AddValue("Name", body_name);
				node_body.AddValue("Disabled", body_scan.disabled);
				node_body.AddValue("Map", SCANUtil.integerSerialize(body_scan));
				node_progress.AddNode(node_body);
			}
			node.AddNode(node_progress);
		}

		public void Update()
		{
			if (scan_background) {
				scanFromAllVessels();
				if (HighLogic.LoadedScene == GameScenes.FLIGHT) {
					SCANui.gui_ping(false);
					SCANui.gui_ping_maptraq();
				}
			}
		}

		public class SCANsensor
		{
			public SCANdata.SCANtype sensor;
			public double fov;
			public double min_alt, max_alt, best_alt;

			public bool inRange;
			public bool bestRange;
		}

		public class SCANvessel
		{
			public Guid id;
			public Vessel vessel;
			public Dictionary<SCANdata.SCANtype, SCANsensor> sensors = new Dictionary<SCANdata.SCANtype, SCANsensor>();
            
			public CelestialBody body;
			public double latitude, longitude;
			public int frame;
			public double lastUT;
		}

		public void Resources(CelestialBody b) //Repopulates the master resources list with data from config nodes
		{
			ResourcesList.Clear();
			if (resourceOverlayType == 0 && SCANreflection.ORSFound)
			{
				foreach (ConfigNode node in GameDatabase.Instance.GetConfigNodes("PLANETARY_RESOURCE_DEFINITION"))
				{
					if (node != null)
					{
						bool resourceAdded = false;
						SCANdata.SCANResource resource = null;
						if ((resource = SCANUtil.ORSConfigLoad(node)) == null)
							continue;
						foreach (SCANdata.SCANResource res in ResourcesList)
						{ //Check to see if the resource is already in the list
							if (resource.name == res.name)
							{
								if (resource.body == b.name)
								{
									res.linear = resource.linear;
									res.fullColor = resource.fullColor;
									res.emptyColor = resource.emptyColor;
									res.ORS_Multiplier = resource.ORS_Multiplier;
									res.ORS_Scalar = resource.ORS_Scalar;
									res.ORS_Threshold = resource.ORS_Threshold;
								}
								resourceAdded = true;
								break;
							}
						}
						if (!resourceAdded)
							ResourcesList.Add(resource);
					}
				}
			}
			else if (resourceOverlayType == 1)
			{
				foreach (ConfigNode node in GameDatabase.Instance.GetConfigNodes("KethaneResource"))
				{
					if (node != null)
					{
						string name = node.GetValue("Resource");
						var colorFull = node.GetValue("ColorFull");
						Color full = ConfigNode.ParseColor(colorFull);
						var colorEmpty = node.GetValue("ColorEmpty");
						Color empty = ConfigNode.ParseColor(colorEmpty);
						float max = 1000000f;
						ConfigNode subNode = node.GetNode("Generator");
						if (subNode != null)
						{
							float.TryParse(subNode.GetValue("MaxQuantity"), out max); //Global max quantity
							foreach (ConfigNode bodySubNode in subNode.GetNodes("Body"))
							{
								string body = bodySubNode.GetValue("name");
								if (body == b.name)
								{
									if (bodySubNode.HasValue("MaxQuantity"))
									{
										float.TryParse(bodySubNode.GetValue("MaxQuantity"), out max); //Optional body-specific max quantity
										break;
									}
								}
							}
						}
						SCANdata.SCANtype type = SCANUtil.OverlayResourceType(name, resourceOverlayType);
						ResourcesList.Add(new SCANdata.SCANResource(name, "", full, empty, true, 1d, 1d, 1d, max, type));
					}
				}
			}
			if (ResourcesList.Count == 0)
				globalOverlay = false;
			if (gridSelection > ResourcesList.Count - 1)
				gridSelection = 0;
		}

		internal Dictionary<Guid, SCANvessel> knownVessels = new Dictionary<Guid, SCANvessel>();

		public void registerSensor(Vessel v, SCANdata.SCANtype sensors, double fov, double min_alt, double max_alt, double best_alt) {
			registerSensor(v.id, sensors, fov, min_alt, max_alt, best_alt);
			knownVessels[v.id].vessel = v;
			knownVessels[v.id].latitude = SCANUtil.fixLatShift(v.latitude);
			knownVessels[v.id].longitude = SCANUtil.fixLonShift(v.longitude);
		}

		public void registerSensor(Guid id, SCANdata.SCANtype sensors, double fov, double min_alt, double max_alt, double best_alt) {
			if(!knownVessels.ContainsKey(id)) knownVessels[id] = new SCANvessel();
			SCANvessel sv = knownVessels[id];
			sv.id = id;
			foreach(SCANdata.SCANtype sensor in Enum.GetValues(typeof(SCANdata.SCANtype))) {
				if(SCANUtil.countBits((int)sensor) != 1) continue;
				if((sensor & sensors) == SCANdata.SCANtype.Nothing) continue;
				double this_fov = fov, this_min_alt = min_alt, this_max_alt = max_alt, this_best_alt = best_alt;
				if(this_max_alt <= 0) {
					this_min_alt = 5000;
					this_max_alt = 500000;
					this_best_alt = 200000;
					this_fov = 5;
					if((sensor & SCANdata.SCANtype.AltimetryHiRes) != SCANdata.SCANtype.Nothing) this_fov = 3;
					if((sensor & SCANdata.SCANtype.AnomalyDetail) != SCANdata.SCANtype.Nothing) {
						this_min_alt = 0;
						this_max_alt = 2000;
						this_best_alt = 0;
						this_fov = 1;
					}
				}
				if(!sv.sensors.ContainsKey(sensor)) sv.sensors[sensor] = new SCANsensor();
				SCANsensor s = sv.sensors[sensor];
				s.sensor = sensor;
				s.fov = this_fov;
				s.min_alt = this_min_alt;
				s.max_alt = this_max_alt;
				s.best_alt = this_best_alt;
			}
		}

		public void unregisterSensor(Vessel v, SCANdata.SCANtype sensors) {
			if(!knownVessels.ContainsKey(v.id)) return;
			SCANvessel sv = knownVessels[v.id];
			sv.id = v.id;
			sv.vessel = v;
			foreach(SCANdata.SCANtype sensor in Enum.GetValues(typeof(SCANdata.SCANtype))) {
				if((sensors & sensor) == SCANdata.SCANtype.Nothing) continue;
				if(!sv.sensors.ContainsKey(sensor)) continue;
				sv.sensors.Remove(sensor);
			}
		}

		public bool isVesselKnown(Guid id, SCANdata.SCANtype sensor) {
			if(!knownVessels.ContainsKey(id)) return false;
			SCANdata.SCANtype all = SCANdata.SCANtype.Nothing;
			foreach(SCANdata.SCANtype s in knownVessels[id].sensors.Keys) all |= s;
			return (all & sensor) != SCANdata.SCANtype.Nothing;
		}

		public bool isVesselKnown(Guid id) {
			if(!knownVessels.ContainsKey(id)) return false;
			return knownVessels[id].sensors.Count > 0;
		}

		public bool isVesselKnown(Vessel v) {
			if(v.vesselType == VesselType.Debris) return false;
			return isVesselKnown(v.id);
		}

		public SCANsensor getSensorStatus(Vessel v, SCANdata.SCANtype sensor) {
			if(!knownVessels.ContainsKey(v.id)) return null;
			if(!knownVessels[v.id].sensors.ContainsKey(sensor)) return null;
			return knownVessels[v.id].sensors[sensor];
		}

		public SCANdata.SCANtype activeSensorsOnVessel(Guid id) {
			if(!knownVessels.ContainsKey(id)) return SCANdata.SCANtype.Nothing;
			SCANdata.SCANtype sensors = SCANdata.SCANtype.Nothing;
			foreach(SCANdata.SCANtype s in knownVessels[id].sensors.Keys) sensors |= s;
			return sensors;
		}

		private int i = 0;
		protected static int last_scan_frame;
		protected static float last_scan_time;
		protected static double scan_UT;
		public static int activeSensors, activeVessels;
		private static int currentActiveSensor, currentActiveVessel;
		public void scanFromAllVessels() {
			if(Time.realtimeSinceStartup - last_scan_time < 1 && Time.realtimeSinceStartup > last_scan_time) return;
			if(last_scan_frame == Time.frameCount) return;
			last_scan_frame = Time.frameCount;
			last_scan_time = Time.realtimeSinceStartup;
			scan_UT = Planetarium.GetUniversalTime();
			currentActiveSensor = 0;
			currentActiveVessel = 0;
			actualPasses = 0;
			maxRes = 0;
            SCANdata bdata = SCANUtil.getData (FlightGlobals.Bodies[i]); //Update coverage for planets one at a time, rather than all together
            bdata.updateCoverage();
            i++;
            if (i >= FlightGlobals.Bodies.Count) i = 0;
			foreach(Vessel v in FlightGlobals.Vessels) {
				if(!knownVessels.ContainsKey(v.id)) continue;
				SCANvessel vessel = knownVessels[v.id];
				SCANdata data = SCANUtil.getData(v.mainBody);
				vessel.vessel = v;
			
				if(!data.disabled) {
					if(v.mainBody == FlightGlobals.currentMainBody || scan_background) {
						if(isVesselKnown(v)) {
							doScanPass(knownVessels[v.id], scan_UT, scan_UT, vessel.lastUT, vessel.latitude, vessel.longitude);
							++currentActiveVessel;
							currentActiveSensor += knownVessels[v.id].sensors.Count;
						}
					}
				}

				vessel.body = v.mainBody;
				vessel.frame = Time.frameCount;
				vessel.lastUT = scan_UT;
				vessel.latitude = SCANUtil.fixLatShift(v.latitude);
				vessel.longitude = SCANUtil.fixLonShift(v.longitude);
			}
			activeVessels = currentActiveVessel;
			activeSensors = currentActiveSensor;
		}

		public int actualPasses, maxRes;
		protected static Queue<double> scanQueue;
		protected void doScanPass(SCANvessel vessel, double UT, double startUT, double lastUT, double llat, double llon) {
			Vessel v = vessel.vessel;
			SCANdata data = SCANUtil.getData(v.mainBody);
			double soi_radius = v.mainBody.sphereOfInfluence - v.mainBody.Radius;
			double alt = v.altitude, lat = SCANUtil.fixLatShift(v.latitude), lon = SCANUtil.fixLonShift(v.longitude);
			double res = 0;
			Orbit o = v.orbit;
			bool uncovered;

			if(scanQueue == null) scanQueue = new Queue<double>();
			if(scanQueue.Count != 0) scanQueue.Clear();

		loop: // don't look at me like that, I just unrolled the recursion
			if(res > 0) {
				if(double.IsNaN(UT)) goto dequeue;
				if(o.ObT <= 0) goto dequeue;
				if(double.IsNaN(o.getObtAtUT(UT))) goto dequeue;
				Vector3d pos = o.getPositionAtUT(UT);
				double rotation = 0;
				if(v.mainBody.rotates) {
					rotation = (360 * ((UT - scan_UT) / v.mainBody.rotationPeriod)) % 360;
				}
				alt = v.mainBody.GetAltitude(pos);
				lat = SCANUtil.fixLatShift(v.mainBody.GetLatitude(pos));
				lon = SCANUtil.fixLonShift(v.mainBody.GetLongitude(pos) - rotation);
				if(alt < 0) alt = 0;
				if(res > maxRes) maxRes = (int)res;
			} else {
				alt = v.heightFromTerrain;
				if(alt < 0) alt = v.altitude;
			}

			if(Math.Abs(lat - llat) < 1 && Math.Abs(lon - llon) < 1 && res > 0) goto dequeue;
			actualPasses++;

			uncovered = res <= 0;
			foreach(SCANsensor sensor in knownVessels[v.id].sensors.Values) {
				if(res <= 0) {
					if(data.getCoverage(sensor.sensor) > 0) uncovered = false;
				}

				sensor.inRange = false;
				sensor.bestRange = false;
				if(alt < sensor.min_alt) continue;
				if(alt > Math.Min(sensor.max_alt, soi_radius)) continue;
				sensor.inRange = true;

				double fov = sensor.fov;
				double ba = Math.Min(sensor.best_alt, soi_radius);
				if(alt < ba) fov = (alt / ba) * fov;
				else sensor.bestRange = true;

				double surfscale = 600000d/v.mainBody.Radius;
				if(surfscale < 1) surfscale = 1;
				surfscale = Math.Sqrt(surfscale);
				fov *= surfscale;
				if(fov > 20) fov = 20;

				int f = (int)Math.Truncate(fov);
				int f1 = f + (int)Math.Round(fov - f);
				
				double clampLat;
				double clampLon;
				for(int x=-f; x<=f1; ++x) {
					clampLon = lon + x;	// longitude does not need clamping
					/*if (clampLon < 0  ) clampLon = 0; */
					/*if (clampLon > 360) clampLon = 360; */
					for(int y=-f; y<=f1; ++y) {
						clampLat = lat + y;
						if (clampLat > 90) clampLat = 90;
						if (clampLat < -90) clampLat = -90;
                        SCANUtil.registerPass(clampLon, clampLat, data, sensor.sensor);
                    }
				}
			}
			if(uncovered) return;
			/* 
			if(v.mainBody == FlightGlobals.currentMainBody) {
				if(res > 0) data.map_small.SetPixel((int)Math.Round(lon) + 180, (int)Math.Round(lat) + 90, palette.magenta);
				else data.map_small.SetPixel((int)Math.Round(lon) + 180, (int)Math.Round(lat) + 90, palette.yellow);
			}
			*/

			if(vessel.lastUT <= 0) return;
			if(vessel.frame <= 0) return;
			if(v.LandedOrSplashed) return;
			if(res >= timeWarpResolution) goto dequeue;

			if(startUT > UT) {
				scanQueue.Enqueue((startUT + UT) / 2);
				scanQueue.Enqueue(startUT);
				scanQueue.Enqueue(UT);
				scanQueue.Enqueue(lat);
				scanQueue.Enqueue(lon);
				scanQueue.Enqueue(res + 1);
			}
			startUT = UT;
			UT = (lastUT + UT) / 2;
			llat = lat;
			llon = lon;
			res = res + 1;
			goto loop;
		
		dequeue:
			if(scanQueue.Count <= 0) return;
			UT = scanQueue.Dequeue();
			startUT = scanQueue.Dequeue();
			lastUT = scanQueue.Dequeue();
			llat = scanQueue.Dequeue();
			llon = scanQueue.Dequeue();
			res = scanQueue.Dequeue();
			goto loop;
		}


		#region Unused code

		//public static int countBits(int i) {
		//    int count;
		//    for(count=0; i!=0; ++count) i &= (i - 1);
		//    return count;
		//}

		//**** Most commented code moved to SCANUtil****
		//public ScienceData getAvailableScience(Vessel v, SCANdata.SCANtype sensor, bool notZero) {
		//    SCANdata data = getData(v.mainBody);
		//    ScienceData sd = null;
		//    ScienceExperiment se = null;
		//    ScienceSubject su = null;
		//    bool found = false;
		//    string id = null;
		//    double coverage = 0f;

		//    if(v.mainBody.pqsController != null) {
		//        if(!found && (sensor & SCANdata.SCANtype.AltimetryLoRes) != SCANdata.SCANtype.Nothing) {
		//            found = true;
		//            id = "SCANsatAltimetryLoRes";
		//            coverage = data.getCoveragePercentage(SCANdata.SCANtype.AltimetryLoRes);
		//        }
		//        if(!found && (sensor & SCANdata.SCANtype.AltimetryHiRes) != SCANdata.SCANtype.Nothing) {
		//            found = true;
		//            id = "SCANsatAltimetryHiRes";
		//            coverage = data.getCoveragePercentage(SCANdata.SCANtype.AltimetryHiRes);
		//        }
		//    }
		//    if(v.mainBody.BiomeMap != null) {
		//        if(!found && (sensor & SCANdata.SCANtype.Biome) != SCANdata.SCANtype.Nothing) {
		//            found = true;
		//            id = "SCANsatBiomeAnomaly";
		//            coverage = data.getCoveragePercentage(SCANdata.SCANtype.Biome); //This should only really check biomes, and it screws up with the changes made to getCoveragePercentage
		//        }
		//    }

		//    if(!found) return null;

		//    se = ResearchAndDevelopment.GetExperiment(id);
		//    if(se == null) return null;

		//    su = ResearchAndDevelopment.GetExperimentSubject(se, ExperimentSituations.InSpaceHigh, v.mainBody, "surface");
		//    if(su == null) return null;

		//    print("[SCANsat] coverage " + coverage.ToString("F1") + ", science cap " + su.scienceCap.ToString("F1") + ", subject value " + su.subjectValue.ToString("F2") + ", science value " + su.scientificValue.ToString("F2") + ", science " + su.science.ToString("F2"));
		//    su.scientificValue = 1;

		//    float science = (float)coverage;
		//    if(science > 95) science = 100;
		//    if(science < 30) science = 0;
		//    science = science / 100f;
		//    science = Mathf.Max(0, (science * su.scienceCap) - su.science);

		//    print("[SCANsat] remaining science: " + science.ToString("F1") + ", base = " + (se.baseValue).ToString("F1"));

		//    science /= Mathf.Max(0.1f, su.scientificValue);
		//    science /= su.subjectValue;

		//    print("[SCANsat] result = " + science.ToString("F2"));

		//    if(notZero && science <= 0) science = 0.00001f;

		//    sd = new ScienceData(science * su.dataScale, 1f, 0f, su.id, se.experimentTitle + " of " + v.mainBody.theName);
		//    su.title = sd.title;
		//    return sd;
		//}

		//public SCANdata getData(string name) {
		//    if(!SCANUtil.body_data.ContainsKey(name)) {
		//        SCANUtil.body_data[name] = new SCANdata();
		//        SCANUtil.body_data[name].resetImages();
		//    }
		//    return SCANUtil.body_data[name];
		//}

		//public SCANdata getData(CelestialBody body) {
		//    SCANdata data = getData(body.bodyName);
		//    data.body = body;
		//    return data;
		//}

		//public void registerPass(CelestialBody body, float lon, float lat, SCANdata.SCANtype type) {
		//    getData(body).registerPass(lon, lat, type);
		//}

		//public double fixLatitude(double lat) {
		//    return (lat + 90 + 180) % 180 - 90;
		//}

		//public double fixLongitude(double lon) {
		//    return (lon + 180 + 360) % 360 - 180;
		//}

		//****Unused****//
		//protected static HashSet<string> disabledBodies = new HashSet<string>();
		//public static bool bodyIsDisabled(string name) {
		//    return disabledBodies.Contains(name);
		//}

		//public static void setBodyDisabled(string name, bool disabled) {
		//    if(disabled) disabledBodies.Add(name);
		//    else disabledBodies.Remove(name);
		//}

		#endregion

	}
}
